packages/core/utils/lib/async.js

```diff
@@ -0,0 +1,69 @@
+'use strict';
+
+function pipeAsync(...methods) {
+  return async (data) => {
+    let res = data;
+
+    for (const method of methods) {
+      res = await method(res);
+    }
+
+    return res;
+  };
+}
+
+/**
+ * Map function callback.
+ * @callback mapAsyncCallback
+ * @param {*} value
+ * @param {number} index
+ */
+/**
+ * Async iteration over an array of promises
+ * @param {promise<*>[]} promiseArray
+ * @returns {function(callback: mapAsyncCallback): promise<*[]>}
+ */
+function mapAsync(promiseArray) {
+  /**
+   * @param {mapAsyncCallback} callback
+   * @returns promise<*[]>
+   */
+  return (callback) => {
+    const transformedPromiseArray = promiseArray.map(async (promiseValue, index) => {
+      const value = await promiseValue;
+      return callback(value, index);
+    });
+    return Promise.all(transformedPromiseArray);
```

Comment:
I think there might be a misunderstanding here. 

We already do Promise.all(array.map(async () => {}}) directly in the codebase whenever we do concurrent execution.

The issue was when you want sequential execution of a map you endup doing for loops so you are not solving that usecase here

Reply:
Bluebird approach looks interesting, being able to say "I want 3 max concurrent executions at a time" or "infinity" or "1".
It let the choice to the function caller to specify the appropriate way to handle the promise array.